core/src/main/java/org/apache/iceberg/io/datafile/WriteBuilder.java

```diff
@@ -0,0 +1,340 @@
+   * and {@link #metricsConfig(MetricsConfig)}.
+   */
+  public WriteBuilder<A, E> forTable(Table table) {
+    appenderBuilder.forTable(table);
+    return this;
+  }
+
+  /** Set the file schema. */
+  public WriteBuilder<A, E> schema(Schema newSchema) {
+    appenderBuilder.schema(newSchema);
+    return this;
+  }
+
+  /**
+   * Sets the engine specific schema for the input. Used by the {@link
+   * AppenderBuilder#build(AppenderBuilder.WriteMode)} to configure the engine specific converters.
+   */
+  public WriteBuilder<A, E> engineSchema(E engineSchema) {
+    appenderBuilder.engineSchema(engineSchema);
+    return this;
+  }
+
+  /** Set the file schema's root name. */
+  public WriteBuilder<A, E> named(String newName) {
+    appenderBuilder.named(newName);
+    return this;
+  }
+
+  /**
+   * Set a writer configuration property.
+   *
+   * <p>Write configuration affects writer behavior. To add file metadata properties, use {@link
+   * #meta(String, String)} or {@link #meta(Map)}.
+   *
+   * @param property a writer config property name
+   * @param value config value
+   * @return this for method chaining
+   */
+  public WriteBuilder<A, E> set(String property, String value) {
+    appenderBuilder.set(property, value);
+    return this;
+  }
+
+  /**
+   * Adds the new properties to the writer configuration.
+   *
+   * <p>Write configuration affects writer behavior. To add file metadata properties, use {@link
+   * #meta(String, String)} or {@link #meta(Map)}.
+   *
+   * @param properties a map of writer config properties
+   * @return this for method chaining
+   */
+  public WriteBuilder<A, E> set(Map<String, String> properties) {
+    properties.forEach(appenderBuilder::set);
+    return this;
+  }
+
+  /**
+   * Set a file metadata property.
+   *
+   * <p>Metadata properties are written into file metadata. To alter a writer configuration
+   * property, use {@link #set(String, String)} or {@link #set(Map)}.
+   *
+   * @param property a file metadata property name
+   * @param value config value
+   * @return this for method chaining
+   */
+  public WriteBuilder<A, E> meta(String property, String value) {
+    appenderBuilder.meta(property, value);
+    return this;
+  }
+
+  /**
+   * Add the new properties to file metadata.
+   *
+   * <p>Metadata properties are written into file metadata. To alter a writer configuration
+   * property, use {@link #set(String, String)}.
+   *
+   * @param properties a map of file metadata properties
+   * @return this for method chaining
+   */
+  public WriteBuilder<A, E> meta(Map<String, String> properties) {
+    properties.forEach(appenderBuilder::meta);
+    return this;
+  }
+
+  /** Sets the metrics configuration used for collecting column metrics for the created file. */
+  public WriteBuilder<A, E> metricsConfig(MetricsConfig newMetricsConfig) {
+    appenderBuilder.metricsConfig(newMetricsConfig);
```

Comment:
Is this ever customized? Can we set it automatically from table properties?

Reply:
The current GenericAppenderFactory customizes it for backward compatibility reasons.
I'm not sure requiring a full table object as a parameter is better anyways.. (I just removed based on your suggestion 😄)